pkg/cover/backend/modules.go

```diff
@@ -47,9 +47,20 @@ func discoverModulesLinux(dirs []string, hostModules []host.KernelModule) ([]*Mo
 			continue
 		}
 		log.Logf(0, "module %v -> %v", mod.Name, path)
+		// .text section of kernel modules may have non-zero offset.
+		// Make sure to adjust the base address to account for it.
+		offset := uint64(0)
+		file, err := elf.Open(path)
+		if err == nil {
+			s := file.Section(".text")
+			if s != nil {
+				offset = s.SectionHeader.Offset
```

Comment:
I don't have much expertise in elf files/module loading, but it looks a bit strange to me.

That offset (`sh_offset`) is the offset from the beginning of the `.ko` file to the first byte of the section [1].

There's also `sh_addr`, which is the memory address where that first byte of the section must reside. So `/proc/modules` [2] output actually does not point to that `sh_addr`, but rather to the first byte of the memory where the whole`.ko` file is mapped?

[1] https://man7.org/linux/man-pages/man5/elf.5.html

[2]
https://github.com/google/syzkaller/blob/28b24332d95f2f7df44ec7e7a5e0025bcadc6277/pkg/host/machine_info_linux.go#L135-L137

Reply:
Please take another look. I dropped the second commit and will mail it separately.